Add more root finding algorithms

VHLM2001 opened the pull request:
Adds bisection method and a modified Regula Falsi method (Illinois method) to find roots of function.

Bisection method is linear, guarantees a decently fast convergence for when other methods are slow to converge. This method follows by calculating whether midpoint is left or right of the root and updating bound on the same side to be the midpoint.

The Illinois method addresses the degenerate case of the Regula Falsi method in which one bound becomes fixed in the iterations leading to extremely slow convergence. This is done by checking if the approximated root calculated in the current iteration is on the same side of the true root as the previous iteration. When this occurs the value of the root function at the bound of the other side is halved.

VHLM2001 committed:
Bisection Method

VHLM2001 committed:
Illinois Method

sethrj commented on src/corecel/math/BisectionRootFinder.hh:
```diff
@@ -0,0 +1,113 @@
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+#include "detail/MathImpl.hh"
```
I suspect this got pulled in by vscode, nothing you use is in there...

sethrj commented on src/corecel/math/BisectionRootFinder.hh:
```diff
@@ -0,0 +1,113 @@
+    do
+    {
+        // Estimate root and update value
+        root = 0.5 * (left + right);
```
```suggestion
        root = real_type(0.5) * (left + right);
```
Otherwise you're doing a float->double->float conversion potentially

sethrj commented on src/corecel/math/IllinoisRootFinder.hh:
```diff
@@ -0,0 +1,122 @@
+ * Perform Regula Falsi (see RegulaFalsi for more details) iterations given a
+ * root function \em func and tolerance \em tol using the Illinois method.
+ *
+ * Illonois method modifies the standard approach by comparing the sign of
```
```suggestion
 * Illinois method modifies the standard approach by comparing the sign of
```

sethrj commented on src/corecel/math/IllinoisRootFinder.hh:
```diff
@@ -0,0 +1,122 @@
+ * the bound on the other side is halved.
+ */
+template<class F>
+class Illinois
```
```suggestion
class IllinoisRootFinder
```

sethrj commented on src/corecel/math/IllinoisRootFinder.hh:
```diff
@@ -0,0 +1,122 @@
+        {
+            left = root;
+            f_left = f_root;
+            if (side == -1)
```
This looks like it should be a boolean enum rather than a real value.

sethrj commented on src/corecel/math/IllinoisRootFinder.hh:
```diff
@@ -0,0 +1,122 @@
+            f_left = f_root;
+            if (side == -1)
+            {
+                f_right *= 0.5;
```
```suggestion
                f_right *= real_type(0.5);
```
and belowas well

sethrj commented on test/corecel/math/IllinoisRootFinder.test.cc:
```diff
@@ -0,0 +1,96 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020-2024 UT-Battelle, LLC, and other Celeritas developers.
```
Since all these files are new, you can just mark them as 2024 :) They're normally generated automatically with `scripts/dev/celeritas-gen.py`.

sethrj commented on test/corecel/math/RegulaFalsiRootFinder.test.cc:
```diff
@@ -65,29 +65,30 @@ TEST(RegulaFalsi, trigometric)
-    double y = 2.0;
-    double u = 0.0;
-    double v = -1.0;
+    double x = 1.5;
+    double y = 0.5;
+    double u = -0.7071067812;
+    double v = 0.7071067812;
```
While you're at it (see #1331) change these to `real_type`, otherwise you're making the lambdas do work in double precision

sethrj reviewed (changes requested):
Only a few small suggestions. Nice work, and isn't it cool how this modular strategy lets us change the involute solver performance with a single line (in the solver itself)?

VHLM2001 committed:
Inplementing comments

sethrj commented on src/corecel/math/IllinoisRootFinder.hh:
```diff
@@ -0,0 +1,130 @@
+class IllinoisRootFinder
+{
+  public:
+    //! Enum defining side of aproximated root to true root
```
This enum can be private, and I think you can even define it local to the function

VHLM2001 committed:
Assertion for bisection method

however if after 50 iterations its still not converged I think there's another problem

sethrj commented on src/corecel/math/BisectionRootFinder.hh:
```diff
@@ -0,0 +1,113 @@
+    inline CELER_FUNCTION BisectionRootFinder(F&& func, real_type tol);
+
+    // Solve for a root between two points
+    inline real_type operator()(real_type left, real_type right);
```
Missing `CELER_FUNCTION`

sethrj commented on src/corecel/math/IllinoisRootFinder.hh:
```diff
@@ -0,0 +1,130 @@
+    inline CELER_FUNCTION IllinoisRootFinder(F&& func, real_type tol);
+
+    // Solve for a root between two points
+    inline real_type operator()(real_type left, real_type right);
```
Missing `CELER_FUNCTION`

sethrj commented on src/corecel/math/RegulaFalsiRootFinder.hh:
```diff
@@ -37,11 +35,11 @@ namespace celeritas
  */
 template<class F>
-class RegulaFalsi
+class RegulaFalsiRootFinder
```
Oops, I didn't even notice on the first review 😄

sethrj commented on src/corecel/math/RegulaFalsiRootFinder.hh:
```diff
@@ -37,11 +35,11 @@ namespace celeritas
+    inline CELER_FUNCTION RegulaFalsiRootFinder(F&& func, real_type tol);
 
     // Solve for a root between two points
     inline real_type operator()(real_type left, real_type right);
```
Might as well add `CELER_FUNCTION` here as well (you did this in the involute branch)

VHLM2001 committed:
Quick fixes

VHLM2001 committed:
Make all methods have the same maximum iterations.

sethrj reviewed (approved):
Excellent work! Thanks!

sethrj committed:
Merge branch 'develop' into iteration-methods
